### PR TITLE
Chapter8

### PR DESCRIPTION
**Pull request by heimrih**
ready

### CODE REVIEW

**Comment by heimrih** [comment]
ready

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -13,4 +13,5 @@ def extract_locale
   def default_url_options
     {locale: I18n.locale}
   end
+  include SessionsHelper
```
Comment (ngocvt-0484) [inline app/controllers/application_controller.rb]: it should below class ApplicationController
not here

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  def new; end
+
+  def create
+    user = User.find_by(email: params[:session][:email].downcase)
+    if user && user.authenticate(params[:session][:password])
+      log_in user
+      redirect_to user
+    else
+      flash[:danger] = t "login.fail"
+      render "new"
```
Comment (ngocvt-0484) [inline app/controllers/sessions_controller.rb]:
```suggestion
      render :new
```

File: app/helpers/sessions_helper.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+module SessionsHelper
+
```
Comment (ngocvt-0484) [inline app/helpers/sessions_helper.rb]: remove empty line

File: app/helpers/sessions_helper.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  def logged_in?
+    !current_user.nil?
+  end
+
+  def log_out
+    session.delete(:user_id)
+    @current_user = nil
+  end
+
```
Comment (ngocvt-0484) [inline app/helpers/sessions_helper.rb]: remove empty line

File: app/helpers/sessions_helper.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  def current_user
+    if session[:user_id]
+      @current_user ||= User.find_by(id: session[:user_id])
+    end
+  end
+
+  def logged_in?
+    !current_user.nil?
```
Comment (ngocvt-0484) [inline app/helpers/sessions_helper.rb]:
```suggestion
    current_user.present?
```

File: app/helpers/sessions_helper.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
+module SessionsHelper
+
+  def log_in user
+    session[:user_id] = user.id
+  end
+
+  def current_user
+    if session[:user_id]
+      @current_user ||= User.find_by(id: session[:user_id])
```
Comment (ngocvt-0484) [inline app/helpers/sessions_helper.rb:8]:
```suggestion
      @current_user ||= User.find_by(id: session[:user_id]) if session[:user_id]
```

File: app/helpers/sessions_helper.rb [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    end
+  end
+
+  def logged_in?
+    !current_user.nil?
+  end
+
+  def log_out
+    session.delete(:user_id)
```
Comment (ngocvt-0484) [inline app/helpers/sessions_helper.rb]:
```suggestion
    session.delete :user_id
```

**Comment by heimrih** [comment]
ready

File: app/helpers/sessions_helper.rb [hunk 8]
```diff
@@ -0,0 +1,20 @@
+module SessionsHelper
+  def log_in user
+    session[:user_id] = user.id
+  end
+
+  def current_user
+    if session[:user_id]
+      @current_user ||= User.find_by(id: session[:user_id]) if session[:user_id]
```
Comment (manhnt-0960) [inline app/helpers/sessions_helper.rb]:
```suggestion
      @current_user ||= User.find_by(id: session[:user_id])
```

File: app/models/user.rb [hunk 9]
```diff
@@ -15,4 +15,10 @@ class User < ApplicationRecord
   has_secure_password
 
   before_save { email.downcase! }
+
+  def User.digest(string)
```
Comment (manhnt-0960) [inline app/models/user.rb]: declare class method with syntax
```class << self```

File: app/javascript/packs/application.js [hunk 10]
```diff
@@ -7,6 +7,8 @@ require("@rails/ujs").start()
 require("turbolinks").start()
 require("@rails/activestorage").start()
 require("channels")
+require("jquery")
+import "bootstrap"
```
Comment (saobv-2223) [inline app/javascript/packs/application.js]: in js and css files, using ' instead of "

File: app/views/layouts/_header.html.erb [hunk 11]
```diff
@@ -5,6 +5,25 @@
       <ul class="nav navbar-nav navbar-right">
         <li><%= link_to t("global.home"), root_path %></li>
         <li><%= link_to t("global.help"), help_path %></li>
+
+        <% if logged_in? %>
+          <li><%= link_to "Users", "#" %></li>
+          <li class="dropdown">
+            <a href="#" class="dropdown-toggle" data-toggle="dropdown">
+              Account <b class="caret"></b>
+            </a>
```
Comment (saobv-2223) [inline app/views/layouts/_header.html.erb]: using link_to

File: app/views/sessions/new.html.erb [hunk 12]
```diff
@@ -0,0 +1,19 @@
+<% provide :title, t("login.title") %>
+<title><%= t "login.title" %></title>
+<h1><%= t "login.title" %></h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(:session, url: login_path) do |f| %>
```
Comment (saobv-2223) [inline app/views/sessions/new.html.erb]: remove ()

File: app/views/sessions/new.html.erb [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+      <%= f.label :email %>
+      <%= f.email_field :email, class: "form-control" %>
+      <%= f.label :password %>
+      <%= f.password_field :password, class: "form-control" %>
+      <%= f.submit t("login.title"), class: "btn btn-primary" %>
+    <% end %>
+    <p>
+      <%= t("login.newuser") %>
```
Comment (saobv-2223) [inline app/views/sessions/new.html.erb]: remove ()

File: config/routes.rb [hunk 14]
```diff
@@ -1,10 +1,14 @@
 Rails.application.routes.draw do
+  get 'sessions/new'
```
Comment (manhnt-0960) [inline config/routes.rb]: remove this line because it's not in use

Comment (saobv-2223) [inline config/routes.rb]: using "
move this routes to scope block

File: app/helpers/sessions_helper.rb (hunk 8 above)
Comment (heimrih) [reply app/helpers/sessions_helper.rb]: But 
if session[:user_id] 
miss Ngoc suggested to add

**Comment by heimrih** [comment]
ready

File: app/controllers/application_controller.rb [hunk 15]
```diff
@@ -13,4 +15,5 @@ def extract_locale
   def default_url_options
     {locale: I18n.locale}
   end
+
```
Comment (vanvtt-0952) [inline app/controllers/application_controller.rb]: remove this line (18)

File: app/views/layouts/_header.html.erb [hunk 16]
```diff
@@ -5,11 +5,44 @@
...
+        <% else %>
+          <li><%= link_to t("login.title"), login_path %></li>
+        <% end %>
       </ul>
-      <ul>
+<!--      <ul>
         <li><%= link_to t("English"), root_path(locale: :en) %></li>
         <li><%= link_to t("Tiếng Việt"), root_path(locale: :vi) %></li>
-      </ul>
+      </ul>-->
```
Comment (vanvtt-0952) [inline app/views/layouts/_header.html.erb]: plz, remove line 30 to 33

File: app/views/sessions/new.html.erb [hunk 17]
```diff
@@ -0,0 +1,19 @@
+<% provide :title, t("login.title") %>
+<title><%= t "login.title" %></title>
+<h1><%= t "login.title" %></h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for :session, url: login_path do |f| %>
+      <%= f.label :email %>
```
Comment (vanvtt-0952) [inline app/views/sessions/new.html.erb]: label tag alo nessacery use `I18n`
follow link https://guides.rubyonrails.org/i18n.html#translations-for-active-record-models

**Comment by heimrih** [comment]
ready

File: app/helpers/sessions_helper.rb (hunk 6 above)
Comment (ngocvt-0484) [reply app/helpers/sessions_helper.rb:8]: if you using this code, you should delete line 7 and 9 
just only need @current_user ||= User.find_by(id: session[:user_id]) if session[:user_id]

File: app/models/user.rb [hunk 18]
```diff
@@ -15,4 +15,12 @@ class User < ApplicationRecord
   has_secure_password
 
   before_save { email.downcase! }
+
+  class << self
+    def User.digest(string)
```
Comment (ngocvt-0484) [inline app/models/user.rb]: this method inside class self, so dont need User anymore
```suggestion
    def digest string
```

**Comment by heimrih** [comment]
ready

**Comment by vanvtt-0952** [comment]
merged

**Pull request merged**